pkg/controller/cluster/cluster_controller.go

```diff
@@ -81,68 +100,76 @@ func (r *ReconcileCluster) Reconcile(request reconcile.Request) (reconcile.Resul
 			// For additional cleanup logic use finalizers.
 			return reconcile.Result{}, nil
 		}
+
 		// Error reading the object - requeue the request.
 		return reconcile.Result{}, err
 	}
 
-	name := cluster.Name
-	klog.Infof("Running reconcile Cluster for %q", name)
+	// Store Cluster early state to allow patching.
+	patchCluster := client.MergeFrom(cluster.DeepCopy())
 
-	// If object hasn't been deleted and doesn't have a finalizer, add one
-	// Add a finalizer to newly created objects.
-	if cluster.ObjectMeta.DeletionTimestamp.IsZero() {
-		finalizerCount := len(cluster.Finalizers)
-
-		if !util.Contains(cluster.Finalizers, metav1.FinalizerDeleteDependents) {
-			cluster.Finalizers = append(cluster.ObjectMeta.Finalizers, metav1.FinalizerDeleteDependents)
+	// Always issue a Patch for the Cluster object and its status after each reconciliation.
+	// TODO(vincepri): Figure out if we should bubble up the errors from Patch to the controller.
```

Comment:
I think we have to do this as part of this PR

Reply:
I'll follow-up if that's ok, I'll need to update it anyways after controller-runtime fixes the linked issue. Same goes for the machine controller

Reply:
I'd rather have it here if it's not too much work. WDYT @detiber?

Reply:
Done, I made it so it doesn't overwrite the return error if it's already non-nil